packages/grid/test/selection.common.js

```diff
@@ -856,4 +856,133 @@ describe('multi selection column', () => {
+    let itemSelectionSpy, rows, checkboxes;
+
+    async function mouseClick(element) {
+      const { x, y, width, height } = element.getBoundingClientRect();
+      await sendMouse({
+        type: 'click',
+        position: [x + width / 2, y + height / 2].map(Math.floor),
+      });
+    }
+
+    function assertEvent(detail) {
+      expect(itemSelectionSpy).to.be.calledOnce;
+      expect(itemSelectionSpy.args[0][0].detail).to.eql(detail);
+      itemSelectionSpy.resetHistory();
+    }
+
+    beforeEach(async () => {
+      grid = fixtureSync(`
+        <vaadin-grid style="width: 200px; height: 450px;">
+          <vaadin-grid-selection-column></vaadin-grid-selection-column>
+          <vaadin-grid-column path="name"></vaadin-grid-column>
+        </vaadin-grid>
+      `);
+      grid.items = [{ name: 'Item 0' }, { name: 'Item 1' }, { name: 'Item 2' }];
+      await nextRender();
+
+      rows = getRows(grid.$.items);
+      checkboxes = [...grid.querySelectorAll('vaadin-checkbox[aria-label="Select Row"]')];
+
+      itemSelectionSpy = sinon.spy();
+      grid.addEventListener('item-toggle', itemSelectionSpy);
+    });
+
+    afterEach(async () => {
+      await resetMouse();
+    });
+
+    it('should fire the event when toggling an item with click', async () => {
+      await mouseClick(checkboxes[0]);
+      assertEvent({ item: grid.items[0], selected: true, shiftKey: false });
+
+      await mouseClick(checkboxes[0]);
+      assertEvent({ item: grid.items[0], selected: false, shiftKey: false });
+    });
+
+    it('should fire the event when toggling an item with Shift + click', async () => {
+      await sendKeys({ down: 'Shift' });
+      await mouseClick(checkboxes[0]);
+      await sendKeys({ up: 'Shift' });
+      assertEvent({ item: grid.items[0], selected: true, shiftKey: true });
+
+      await sendKeys({ down: 'Shift' });
+      await mouseClick(checkboxes[0]);
+      await sendKeys({ up: 'Shift' });
+      assertEvent({ item: grid.items[0], selected: false, shiftKey: true });
+    });
+
+    it('should fire the event when toggling an item with Space', async () => {
+      checkboxes[0].focus();
+
+      await sendKeys({ press: 'Space' });
+      assertEvent({ item: grid.items[0], selected: true, shiftKey: false });
+
+      await sendKeys({ press: 'Space' });
+      assertEvent({ item: grid.items[0], selected: false, shiftKey: false });
+    });
+
+    it('should fire the event when toggling an item with Shift + Space', async () => {
+      checkboxes[0].focus();
+
+      await sendKeys({ down: 'Shift' });
+      await sendKeys({ press: 'Space' });
+      await sendKeys({ up: 'Shift' });
+      assertEvent({ item: grid.items[0], selected: true, shiftKey: true });
+
+      await sendKeys({ down: 'Shift' });
+      await sendKeys({ press: 'Space' });
+      await sendKeys({ up: 'Shift' });
+      assertEvent({ item: grid.items[0], selected: false, shiftKey: true });
+    });
+
+    describe('autoSelect', () => {
```

Comment:
One small issue I noticed with auto-select:
- Press space on a regular cell that is not in selection column
- Move down some cells with arrow keys
- Press shift
- Press space

There is no range select, because `item-toggle` is fired before `__onGridInteraction`, so the shift key is not registered.

Could track this as a separate issue if this becomes to complex to fix in this PR.

Reply:
Thanks, fixed.